### ISSUE TITLE
[Infrastructure] Switch to a more callback-oriented architecture

### ISSUE THREAD

**Issue by aicioara** [palantir/plottable#2604]
I am not sure if this is a good idea, but it is worth a thought. I see myself writing this sort of code a lot

```typescript
if (this.x() && this.x().scale && this.y() && this.y().scale) {
  // do something
}
```

Not only that but I will still want to do the `// do something`, once the scales are available, so I end up writing it into the `.x()` or `.y()`, which is 1. not pretty 2. opens room for a lot of edge case errors (say I forgot to add the code in the .x(), which is also the same code I write inside the if (exemplified above).

Suggestion is to move to a callback based architecture. We need to figure out what are the best points to add to the Component API, but it would look like

```typescript
Component.onRenderFinish()
Component.offRenderFinish()

Component.onProjectorsReady()
Component.offProjectorsReady()

Component.onAnchor() // exists already
Component.offAnchor() // exists already
```

Comments?

**Comment by aicioara** [comment]
It is also worth bringing this up in the upcoming architecture meeting.

**Comment by aicioara** [comment]
@bluong @jtlan @ztsai RFC

**Comment by danmane** [comment]
Since sometimes setup on a Plottable chart can only occur after the chart has rendered, it would be very nice to have `Plottable.renderTo` take a callback that is called once all the setup and rendering is finished.

I imagine this being distinct from `onRenderFinish` in that it would evaluate only once, even if other events cause the plot to re-render.

**Comment by aicioara** [comment]
that's a very good point. maybe call the API point `onFirstRenderFinish()`

**Comment by drglitch** [comment]
+1 on this. would be great to have `onInitialRenderFinish` hook to be able to grab the SVG and do various things with it (like tooltips).
ideally,it would also be per `XYPlot` descendant as opposed to entire `Table`

A workaround is setting a data attribute at this time, but its hacky.

**Comment by jtlan** [comment]
Looks like `onRender()` has its own issue here https://github.com/palantir/plottable/issues/1755.
